### PR TITLE
Add aarch64 support

### PR DESCRIPTION
**Pull request by meyskens**
This PR adds aarch64/arm64 build files for Debian stretch/jessie and Ubuntu Xenial and newer.
This PR is like #25 (sorry saw it right before sending this in) but uses the golang-backports apt repo till Go builds aarch64 binaries and is not only for Ubuntu Xenial.

I tested running the .deb on Debian Stretch. Any feedback welcome.

### CODE REVIEW

**Comment by douglasmiranda** [comment]
So, I'd like to share my experiences with packaging docker aarch on Pine64 running DietPI.

I tried to `make deb` with the Debian Jessie aarch Dockerfile, but with no success. So I changed the image to `arm64v8/debian:jessie-backports`. For Debian Jessie there is no aarch64 build on stable channel, only on backports.

As you can see here:

https://packages.debian.org/jessie/libseccomp-dev
https://packages.debian.org/jessie-backports/libseccomp-dev

More info:

Documented the steps here:
https://gist.github.com/douglasmiranda/41c582be2e216f09e1c1081a69e66f87

---

As you can see on my steps, I've cloned the /docker-ce repo no /docker-ce-packaging. 

So my doubt is, the code from here will be merged on /docker-ce?

--- 

I hope I could help, please tell if there's a misconception from my part.

**Comment by douglasmiranda** [comment]
There's this too: https://github.com/moby/moby/blob/master/contrib/builder/deb/aarch64/debian-jessie/Dockerfile haven't tried yet.

**Comment by meyskens** [comment]
@douglasmiranda thx (sorry for the delay) will add backports on Jessie

**Comment by douglasmiranda** [comment]
xD Nice!

File: deb/ubuntu-zesty/Dockerfile.aarch64 [hunk 1]
```diff
@@ -0,0 +1,28 @@
+FROM multiarch/ubuntu-core:arm64-yakkety
```
Comment (suihkulokki) [inline deb/ubuntu-zesty/Dockerfile.aarch64]: This for ubuntu zesty, so arm64-yakkety is incorrect image. I think using arm64v8/ubuntu:zesty is would be the best here. arm64v8/ seems to be the closest thing to official debian/ubuntu images and should probably be used elsewhere

File: deb/debian-jessie/Dockerfile.aarch64 [hunk 2]
```diff
@@ -0,0 +1,33 @@
+FROM multiarch/debian-debootstrap:arm64-jessie
+
+# allow replacing httpredir or deb mirror
+ARG APT_MIRROR=deb.debian.org
+RUN sed -ri "s/(httpredir|deb).debian.org/$APT_MIRROR/g" /etc/apt/sources.list
+RUN echo "deb http://$APT_MIRROR/debian jessie-backports main" >> /etc/apt/sources.list
```
Comment (suihkulokki) [inline deb/debian-jessie/Dockerfile.aarch64]: perhaps use arm64v8/debian:jessie-backports ?

**Comment by suihkulokki** [comment]
Generally looks good to me. I think using the [arm64v8](https://hub.docker.com/r/arm64v8/) namespace for base images would be an improvement.

**Comment by meyskens** [comment]
@suihkulokki will do!

**Comment by suihkulokki** [comment]
Looks good to me now :+1:

**Comment by DieterReuter** [comment]
Any progress here? I'd like to get Debian supported.
@meyskens you should at least resolve the conflict in file "detect_alpine_image".

**Comment by vielmetti** [comment]
I'd like to see this as well, echoing @DieterReuter - one of the desired endpoints for this is an arm64 build for the Raspberry Pi 3.

**Comment by justincormack** [comment]
Needs a rebase

**Comment by meyskens** [comment]
Rebase done

File: detect_alpine_image [hunk 3]
```diff
@@ -6,7 +6,9 @@ img="unknown"
 if [ "$arch" = "x86_64" ]; then
     img="alpine"
 elif [ "$arch" = "armv7l" ]; then
-    img="arm32v6/alpine"
+    img="armhf/alpine"
```
Comment (StefanScherer) [inline detect_alpine_image]: I think this PR shouldn't change the behaviour for other platforms. Probably an artifact of other parallel merges. Upstream already contains the changes for armv7l and aarch64.

**Comment by vielmetti** [comment]
It's worth noting that there is work in progress to get a multiarch Alpine image, per https://github.com/gliderlabs/docker-alpine/issues/304 . That's not done yet, but when it is done this will simplify the logic.

File: detect_alpine_image (hunk 3 above)
Comment (meyskens) [reply detect_alpine_image]: Indeed, that was the just solved merge conflict

**Comment by thaJeztah** [comment]
ping @seemethere @andrewhsu @tianon PTAL

File: deb/ubuntu-zesty/Dockerfile.aarch64 [hunk 4]
```diff
@@ -0,0 +1,28 @@
+FROM aarch64/ubuntu:zesty
```
Comment (tianon) [inline deb/ubuntu-zesty/Dockerfile.aarch64]: This one should be `arm64v8/ubuntu` as well. :+1:

File: detect_alpine_image [hunk 5]
```diff
@@ -7,6 +7,8 @@ if [ "$arch" = "x86_64" ]; then
     img="alpine"
 elif [ "$arch" = "armv7l" ]; then
     img="arm32v6/alpine"
+elif [ "$arch" = "aarch64" ]; then
+    img="arm64v8/alpine"
```
Comment (tianon) [inline detect_alpine_image]: This block is already down a few lines :smile:

File: deb/debian-jessie/Dockerfile.aarch64 [hunk 6]
```diff
@@ -0,0 +1,32 @@
+FROM arm64v8/debian:jessie-backports
+
+# allow replacing httpredir or deb mirror
+ARG APT_MIRROR=deb.debian.org
+RUN sed -ri "s/(httpredir|deb).debian.org/$APT_MIRROR/g" /etc/apt/sources.list
+
+RUN apt-get update && apt-get install -y apparmor bash-completion btrfs-tools build-essential cmake curl ca-certificates debhelper dh-apparmor dh-systemd git libapparmor-dev libdevmapper-dev libltdl-dev libseccomp-dev pkg-config vim-common libsystemd-dev gnupg dirmngr --no-install-recommends && rm -rf /var/lib/apt/lists/*
+
+ENV GO_VERSION 1.9
```
Comment (tianon) [inline deb/debian-jessie/Dockerfile.aarch64]: For consistency across arches, this should still be 1.8.3, shouldn't it? (and they should all be updated together?)

Comment (tianon) [reply deb/debian-jessie/Dockerfile.aarch64]: Oh duh, there weren't binary releases of Go for arm64 until 1.9 -- makes more sense. :innocent:

Comment (thaJeztah) [reply deb/debian-jessie/Dockerfile.aarch64]: Oh, that may be an issue still; we were holding off bumping to 1.9 as there were still some issues (possibly not for all architectures though); https://github.com/moby/moby/pull/33892

**Comment by ericcurtin** [comment]
:+1: Personally I'd be hoping to add fedora 27 support as it is creating disk images for lots of boards

https://github.com/docker/docker-install/issues/40

Maybe I can take a peek at your work and get it done sometime!

**Comment by DieterReuter** [comment]
Hmm, I'm not quite sure why this PR is still pending. Do we really have any major concerns to merge it? Or what parts should be fixed to finally accept it? I'd really love to get this in!

**Comment by meyskens** [comment]
@DieterReuter from my understanding it is Go 1.9 which is the first with official arm64 support. If wanted I can always build 1.8.x from source if 1.9 support is far away

**Comment by DieterReuter** [comment]
Yes maybe, but there is already a version for Ubuntu Xenial built witn Go 1.8.5, just see here https://github.com/docker/docker-ce-packaging/blob/master/deb/ubuntu-xenial/Dockerfile.aarch64.

**Comment by DieterReuter** [comment]
@meyskens wait, I think there is a new solution now! All the other .deb builds are now based upon Go v1.8.5 and actually there is a new pre-built Go v1.8.5 tarball available for armv8! Here it is https://redirector.gvt1.com/edgedl/go/go1.8.5.linux-arm64.tar.gz. So you could now use Go v1.8.5 directly. WDYT?

**Comment by meyskens** [comment]
Oh hey that is nice! Changed the versions to 1.8.5

**Comment by douglasmiranda** [comment]
Just a feedback

I'm using Docker aarch64 on Debian Stretch for two weeks, no issues so far.

I've tried the multiarch official images;
and created some of my own;
using docker-compose (in a container);

Running now: samba file server, afp file server (netatalk), boinc.

Details: https://gist.github.com/douglasmiranda/6c78a82b71f626c8deb930bede6d7dfc

No issues, no warnings, only joy xD

**Comment by DieterReuter** [comment]
Ping @thaJeztah. Could you please :pray: verify internally what we have to do in order to get this PR merged in the near future?
IMO this PR seems to be ready to get merged, it works great for me and @douglasmiranda as well.

**Comment by vielmetti** [comment]
reping @seemethere @andrewhsu @tianon PTAL

**Comment by thaJeztah** [comment]
Checking :hugging_face:

**Comment by seemethere** [comment]
Going to squash and merge this in due to the number of commits.

**Comment by seemethere** [comment]
Thank you so much @meyskens and everyone else that participated in this PR, sorry for the delay but happy to get more `aarch64` support up and running for the upcoming release!

**Comment by DieterReuter** [comment]
🎉woot!

**Pull request merged**